File: src/main/python/prefixsum/이모스의장애물경주17611.py
N = int(input())

arr = []
arr_x = []
arr_y = []
for i in range(N):
  x, y = map(int, input().split())
  arr.append([x, y])
  arr_x.append(x)
  arr_y.append(y)

arr_x.append(arr_x[0])
arr_y.append(arr_y[0])

def get_max(array, minimum, diff):
  graph = [0] * (diff + 1)
  for i in range(N):
    if array[i] != array[i+1]:
      low = min(array[i], array[i+1]) - minimum
      high = max(array[i], array[i+1]) - minimum
      graph[low] += 1
      if high < diff + 1:
        graph[high] -= 1

  prefix = [0] * (diff + 1)
  prefix[0] = graph[0]
  for i in range(1, diff + 1):
    prefix[i] = prefix[i-1] + graph[i]

  return max(prefix)

max_y = max(arr_y)
min_y = min(arr_y)
diff_y = max_y - min_y

max_x = max(arr_x)
min_x = min(arr_x)
diff_x = max_x - min_x

h = get_max(arr_y, min_y, diff_y)
v = get_max(arr_x, min_x, diff_x)
print(max(h, v))

File: src/main/java/org/snust/prefixsum/이모스의장애물경주17611.java
package org.snust.prefixsum;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 이모스의장애물경주17611 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        int[] arrX = new int[N + 1];
        int[] arrY = new int[N + 1];

        int maxX = Integer.MIN_VALUE;
        int maxY = Integer.MIN_VALUE;
        int minX = Integer.MAX_VALUE;
        int minY = Integer.MAX_VALUE;

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());

            arrX[i] = x;
            arrY[i] = y;

            maxX = Math.max(maxX, x);
            maxY = Math.max(maxY, y);
            minX = Math.min(minX, x);
            minY = Math.min(minY, y);
        }

        arrX[N] = arrX[0];
        arrY[N] = arrY[0];

        int diffX = maxX - minX;
        int diffY = maxY - minY;

        int result = Math.max(
                getMax(N, arrX, minX, diffX),
                getMax(N, arrY, minY, diffY)
        );

        System.out.println(result);
    }

    private static int getMax(int N, int[] array, int minimum, int difference) {
        int[] graph = new int[difference + 1];

        for (int i = 0; i < N; i++) {
            int low = Math.min(array[i] - minimum, array[i + 1] - minimum);
            int high = Math.max(array[i] - minimum, array[i + 1] - minimum);

            if (array[i] != array[i + 1]) {
                graph[low]++;
                graph[high]--;
            }
        }

        int[] prefix = new int[difference + 1];
        for (int i = 0; i < graph.length; i++) {
            prefix[i] = (i == 0) ? graph[i] : prefix[i-1] + graph[i];
        }

        return java.util.Arrays.stream(prefix)
                .max()
                .orElse(0);
    }
}